shared_client_test.go

```diff
@@ -0,0 +1,514 @@
+//go:build go1.18
+// +build go1.18
+
+package dns
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+var (
+	clients = NewSharedClients()
+)
+
+func TestSharedClientSync(t *testing.T) {
+	HandleFunc("miek.nl.", HelloServer)
+	defer HandleRemove("miek.nl.")
+
+	s, addrstr, _, err := RunLocalUDPServer(":0")
+	if err != nil {
+		t.Fatalf("unable to run test server: %v", err)
+	}
+	defer s.Shutdown()
+
+	m := new(Msg)
+	m.SetQuestion("miek.nl.", TypeSOA)
+
+	c, closer := clients.GetSharedClient("client-key", new(Client), addrstr)
+	defer closer()
+	r, _, err := c.ExchangeShared(m)
+	if err != nil {
+		t.Fatalf("failed to exchange: %v", err)
+	}
+	if r == nil {
+		t.Fatal("response is nil")
+	}
+	if r.Rcode != RcodeSuccess {
+		t.Errorf("failed to get an valid answer\n%v", r)
+	}
+	// And now another ExchangeAsync on the same shared client
+	r, _, err = c.ExchangeShared(m)
+	if err != nil {
+		t.Errorf("failed to exchange: %v", err)
+	}
+	if r == nil || r.Rcode != RcodeSuccess {
+		t.Errorf("failed to get an valid answer\n%v", r)
+	}
+
+	// Now get the shared client again and make sure it is still the same client
+	c2, closer2 := clients.GetSharedClient("client-key", new(Client), addrstr)
+	defer closer2()
+
+	if c2 != c {
+		t.Fatal("client not really shared")
+	}
+	m.Id = uint16(42)
+	r, _, err = c2.ExchangeShared(m)
+	if err != nil {
+		t.Errorf("failed to exchange: %v", err)
+	}
+	if r == nil || r.Rcode != RcodeSuccess {
+		t.Errorf("failed to get an valid answer\n%v", r)
+	}
+}
+
+func TestSharedClientConcurrentSync(t *testing.T) {
+	HandleFunc("miek.nl.", HelloServer)
+	defer HandleRemove("miek.nl.")
+
+	s, addrstr, _, err := RunLocalUDPServer(":0")
+	if err != nil {
+		t.Fatalf("unable to run test server: %v", err)
+	}
+	defer s.Shutdown()
+
+	conf := &Client{
+		Timeout: time.Second,
+	}
+
+	errors := sync.Map{}
+	var wg sync.WaitGroup
+	f1 := func(id uint16) {
+		defer wg.Done()
+
+		m := new(Msg)
+		m.SetQuestion("miek.nl.", TypeSOA)
+		m.Id = id
+
+		c, closer := clients.GetSharedClient("concurrent-client", conf, addrstr)
+		defer closer()
+		r, _, err := c.ExchangeShared(m)
+		if err != nil {
+			errors.Store(id, fmt.Errorf("failed to exchange: %v", err))
+		} else {
+			if r == nil {
+				errors.Store(id, fmt.Errorf("response is nil"))
+			} else {
+				if r.Id != id {
+					errors.Store(id, fmt.Errorf("incorrect id (%d != %d)", r.Id, id))
+				}
+				if r.Rcode != RcodeSuccess {
+					errors.Store(id, fmt.Errorf("failed to get an valid answer\n%v", r))
+				}
+			}
+		}
```

Comment:
Early return makes this more readable:

```suggestion
		if err != nil {
			errors.Store(id, fmt.Errorf("failed to exchange: %v", err))
			return
		}
		if r == nil {
			errors.Store(id, fmt.Errorf("response is nil"))
			return
		}
		if r.Id != id {
			errors.Store(id, fmt.Errorf("incorrect id (%d != %d)", r.Id, id))
			return
		}
		if r.Rcode != RcodeSuccess {
			errors.Store(id, fmt.Errorf("failed to get an valid answer\n%v", r))
		}
```

Reply:
Nice, adopted!